### ISSUE TITLE
Simplifying Eliom service interface

### ISSUE THREAD

**Issue by balat** [ocsigen/eliom#209]
There is a simplification that I have been wanting to do for years in the interface of Eliom_service and that is mostly possible now (I think) thanks to GADTs.

The idea would be mainly to have only one function to create a service, with some parameters to decide the kind of identification we want (path, parameter or both) and the method (GET/POST).

The main difficulty is that these different kinds of services do not take exactly the same parameters.
For example POST services take a GET service as fallback, etc.

Any idea welcome :)

**Comment by vasilisp** [comment]
A solution would be to make the fallback part of the "identification" data structure (for the applicable cases). The types become scary because the fallback belongs in the type `service` (which has 10 type parameters), but we will simplify things (*somehow*).

**Comment by vasilisp** [comment]
It took quite a while, but the `service` branch now has an all-encompassing [`service` function](https://github.com/ocsigen/eliom/blob/544427efb6c707ea9fc7fbf4481c616219ce0ed9/src/lib/eliom_service_sigs.shared.mli#L104).

**Comment by Drup** [comment]
```ocaml
    meth:('m, 'gp, 'gn, 'pp, 'pn, 'tipo, 'mf, 'gp_) meth ->
    id:('att, 'co, 'mf, 'rt, 'gp_) id ->
    unit ->
    ('gp, 'pp, 'm, 'att, 'co, non_ext, reg, 'tipo, 'gn, 'pn, 'rt) service
```

Blurb, that's not very digestible. ;)

**Comment by vasilisp** [comment]
`meth` should be comprehensible: the method is one of `Get`, `Post`, `Put`, and `Delete`, and all these come with `GET` parameters, except `POST` which also comes with `POST` parameters. 

`id` isn't complicated either: a service is either located on a path `p` (`Path p`), or it sits on top of a `fallback` service (`Overlay fallback`), or it answers on all URLs thanks to service identification magic (`Global`).

So basically there is almost no redundancy in the information you need to provide.

The type parameters were there before, and they can't really go away :).

**Comment by Drup** [comment]
Oh, I agree that the *constructors* are fine, it's the *types* that are quite scary. Unfortunately, OCaml users are used to be directed by types. But I'm sure documentation will be enough. :)

**Comment by vasilisp** [comment]
I understand the objection, and I am not even sure that documentation is a sufficient remedy.
However, there are good reasons why all these parameters are there. Let me write down my view of them in case this triggers any ideas :).

I hope to clean up `'gp`, `'pp`, `'gn`, `'pn` (GET/POST parameters/names) with a pass over `Eliom_parameter`. Hopefully the GET/POST "pairs" of type params can be merged.

 `'tipo` (which stands for with or without suffix, and I have no clue about the name) bothers me, but I have to understand the suffix parameter stuff before I do anything about it.

`'rt` ("return") is there because a service is a general mechanism for things that return disparate things (OCaml values, text, HTML nodes, ...).

`'m` is the HTTP method and `'mf` is the HTTP method of the fallback. That is a quite critical piece of info...

`'gp_` is a hack that encodes the constraint `id = Path _ || 'gp = unit`, thus enforcing unit GET params in _some_ cases. Without this the relationship between the args of the service and of its fallback becomes impossible to encode :(.

The rest are binary flags for various "properties" of the services (is attached, is coservice, is registrable, is external). They could be replaced by runtime assertions plus lots of documentation, but IMHO it would be a backwards step to not reflect them on the type system.

**Comment by Drup** [comment]
Ok, let me clarify: it was *not* an objection, just a (slightly humorous) remark. Any reduction of the number of variable would be welcome, though.

**Comment by vasilisp** [comment]
The types remain an issue, but the API unification did happen (#293), so closing.
